### PR TITLE
[F2F-944] Handle govuk_signin_journey_id in the Return Service

### PR DESCRIPTION
**Pull request by ViveakV**
### What changed

**In Scope**

- Add logInsights to the 3 lambdas within IPR 

- Add BE logic to the postEvent lambda to extract the govuk_signin_journey_id from the IPV_F2F_CRI_VC_CONSUMED event and save this to the sessionTable

- Include this govuk_signin_journey_id in:

    - the logs for the session lambda
    - the logs for the govNotify lambda
    - the IPR_RESULT_NOTIFICATION_EMAILED TxMA event

- For the IPR_USER_REDIRECTED event:

    - assign the newly generated journey_id to govuk_signin_journey_id
    - assign the previous journey_id to extensions.previous_govuk_signin_journey_id

**AUTH_IPV_AUTHORISATION_REQUESTED**
```
{
   "event_id":"588f4a66-f75a-4728-9f7b-8afd865c233c",
   "client_id":"RqFZ83csmS4Mi4Y7s7ohD9-ekwU",
   "clientLandingPageUrl":"https://www.gov.uk/request-copy-criminal-record",
   "event_name":"AUTH_IPV_AUTHORISATION_REQUESTED",
   "rp_name":"replay",
   "timestamp":1681902001,
   "timestamp_formatted":"2023-04-19T11:00:01.000Z",
   "user":{
      "user_id":"",
      "email":"viveak.vadivelkarasan@digital.cabinet-office.gov.uk",
      "govuk_signin_journey_id":"UNKNOWN"
   }
}
```
image

image

```
{
   "event_id":"588f4a66-f75a-4728-9f7b-8afd865c233d",
   "client_id":"RqFZ83csmS4Mi4Y7s7ohD9-ekwU",
   "clientLandingPageUrl":"https://www.gov.uk/request-copy-criminal-record",
   "event_name":"F2F_YOTI_START",
   "rp_name":"replay",
   "timestamp":1681902003,
   "timestamp_formatted":"2023-04-19T11:00:01.000Z",
   "user":{
      "user_id":""
   }
}
```

image

image

```
{
   "event_id":"588f4a66-f75a-4728-9f7b-8afd865c233e",
   "client_id":"RqFZ83csmS4Mi4Y7s7ohD9-ekwU",
   "clientLandingPageUrl":"https://www.gov.uk/request-copy-criminal-record",
   "event_name":"IPV_F2F_CRI_VC_CONSUMED",
   "rp_name":"replay",
   "timestamp":1681902005,
   "timestamp_formatted":"2023-04-19T11:00:01.000Z",
   "user":{
      "user_id":"",
      "govuk_signin_journey_id":"VIVEEAKTESTING"
   },
   "restricted":{
      "nameParts":[
         {
            "type":"GivenName",
            "value":"ANGELA"
         },
         {
            "type":"GivenName",
            "value":"ZOE"
         },
         {
            "type":"FamilyName",
            "value":"UK SPECIMEN"
         }
      ]
   }
}
```
image

image

image

image

image

### CODE REVIEW

File: src/models/SessionReturnRecord.ts [hunk 1]
```diff
@@ -9,6 +10,9 @@ export interface NamePart {
 export class SessionReturnRecord {
 	constructor(data: ReturnSQSEvent, expiresOn: number) {
 		this.userId = data.user.user_id;
+		if (data.user.govuk_signin_journey_id && (data.user.govuk_signin_journey_id).toLowerCase() !== "unknown") {
+			this.clientSessionId = data.user.govuk_signin_journey_id;
```
Comment (richa-misra) [inline src/models/SessionReturnRecord.ts:14]: this is only needed to be extracted from VC_CONSUMED event

File: src/services/PostEventProcessor.ts [hunk 2]
```diff
@@ -147,6 +147,13 @@ export class PostEventProcessor {
 				throw new AppError(HttpCodesEnum.SERVER_ERROR, "Missing event config");
 			}
 
+			if (returnRecord.clientSessionId) {
+				updateExpression += ", clientSessionId = :clientSessionId";
```
Comment (richa-misra) [inline src/services/PostEventProcessor.ts:151]: adding it here would overwrite it for all event types

File: src/services/SessionProcessor.ts [hunk 3]
```diff
@@ -125,11 +125,14 @@ export class SessionProcessor {
 			const sub = jwtIdTokenPayload.sub!;
 			try {
 				session = await iprService.getSessionBySub(sub);
-				this.logger.debug("Session retrieved from session store");
 				if (!session) {
 					this.logger.error("No session event found for this userId");
 					return new Response(HttpCodesEnum.UNAUTHORIZED, "No session event found for this userId");
 				}
+				this.logger.appendKeys({
+					govuk_signin_journey_id: session.clientSessionId,
```
Comment (richa-misra) [inline src/services/SessionProcessor.ts:133]: I think this `govuk_signin_journey_id` would be a new one and needs to be extracted from the JWT

File: src/models/SessionEvent.ts [hunk 4]
```diff
@@ -38,6 +39,10 @@ export class SessionEvent {
 	@IsNotEmpty()
 	userId!: string;
 
+	@IsString()
+	@IsNotEmpty()
```
Comment (richa-misra) [inline src/models/SessionEvent.ts:43]: its possible that clientSessionId is not present

Comment (Stelios-CB) [reply src/models/SessionEvent.ts]: Resolved https://github.com/alphagov/di-ipvreturn-api/pull/56/commits/3bef98ecba1b71c45d234cc1cd1f55168da8e20b

**Review by richa-misra — APPROVED** [review]

**Pull request merged**